# ANSWERS 4.9 PART 1

# QUESTION 3: IMPORT LIBRARIES

In [7]:
# Import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [8]:
# Creating path asa shortcut to access the Instacart Basket data

path = r'C:\Users\carol\Documents\20.06 Instacart Basket Analysis'

In [9]:
# Importing data from customers.csv 

customers = pd.read_csv(os.path.join(path,'02 Data','Original Data','customers.csv'))

# QUESTION 4. WRANGLE DATA

In [10]:
# Review customers.csv heading 

customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [11]:
# Review descriptive analisys to identify any annomalies ( removing user ID that as a numeric number is not usefull for thr description)

customers.drop(columns=['user_id']).describe

<bound method NDFrame.describe of        First Name    Surnam  Gender           STATE  Age date_joined  \
0         Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1        Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        Michelle     Hicks  Female            Iowa   40    1/1/2017   
4             Ann   Gilmore  Female        Maryland   26    1/1/2017   
...           ...       ...     ...             ...  ...         ...   
206204       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207       Rose   Rollins  Female      California   27    4/1/2020   
206208    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                

**NOTE** Age ,number of dependants and income (mean, median and max) seems aligment. Although the max amount of income is far from the mean and median.

In [12]:
# Aligning column names 

customers.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [13]:
# Rename Surnam as last_name

customers.rename(columns = {'Surnam' : 'last_name'}, inplace = True)

In [14]:
# Rename STATE as state

customers.rename(columns = {'STATE' : 'state'}, inplace = True)

In [15]:
# Rename Age as age

customers.rename(columns = {'Age' : 'age'}, inplace = True)

In [16]:
# Rename Genre as gender

customers.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [17]:
# Rename fam_status as marital_status

customers.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [18]:
# checking new column names

customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants marital_status  income  
0             3        married  165665  
1             0         single   59285  
2             2        married   99568  
3             0         single   42049  
4             1        married   40374

**NOTE**6 values were modified to align column name format. I am not dropping any columns. I only completed the descriptive analysis without the user ID but the column will remain as part of the DF

# QUESTION 5. DATA QUALITY

In [19]:
# Review any missing values

customers.isnull().sum()

user_id               0
first_name        11259
last_name             0
gender                0
state                 0
age                   0
date_joined           0
n_dependants          0
marital_status        0
income                0
dtype: int64

**NOTE**  There seems to be 11,259 null values under column first_names

In [20]:
# Reviwing any duplicates

df_dups = customers[customers.duplicated()]

In [21]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependants, marital_status, income]
Index: []

**NOTE** no duplicated data were found

In [22]:
# Checking data types: 

customers.dtypes

user_id            int64
first_name        object
last_name         object
gender            object
state             object
age                int64
date_joined       object
n_dependants       int64
marital_status    object
income             int64
dtype: object

In [23]:
# Checking for any mixed-type columns

for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

C:\Users\carol\AppData\Local\Temp\ipykernel_2188\3951281631.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\carol\AppData\Local\Temp\ipykernel_2188\3951281631.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\carol\AppData\Local\Temp\ipykernel_2188\3951281631.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\carol\AppData\Local\Temp\ipykernel_2188\3951281631.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\carol\A

first_name


C:\Users\carol\AppData\Local\Temp\ipykernel_2188\3951281631.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)


**NOTE**: Column first_name seems to have some mixed type



In [24]:
# As the column contains name changing too a string

customers['first_name'] = customers['first_name'].astype('str')

In [25]:
# Re-Checking for any mixed-type columns

for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)


C:\Users\carol\AppData\Local\Temp\ipykernel_2188\3056519848.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\carol\AppData\Local\Temp\ipykernel_2188\3056519848.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\carol\AppData\Local\Temp\ipykernel_2188\3056519848.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\carol\AppData\Local\Temp\ipykernel_2188\3056519848.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\carol\A

**NOTE**: first_name mix type has been addressed.

# QUESTION 6. COMBINE CUSTOMERS DATA WITH THE PREPARED DATA

In [26]:
# Importing data from ords_prods_grouped 

ords_prods_grouped = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_grouped.pkl'))

In [27]:
# Checking ords_prods_grouped data types

ords_prods_grouped.dtypes

Unnamed: 0                       int64
product_id                       int64
product_name                    object
aisle_id                         int64
department_id                    int64
prices                         float64
order_id                         int64
user_id                          int64
eval_set                        object
order_number                     int64
order_dow                        int64
order_hour_of_day                int64
days_since_prior_order         float64
add_to_cart_order                int64
reordered                        int64
_merge                        category
max_order                        int64
loyalty_flag                    object
mean_product_price             float64
spending_flag                   object
median_days_between_orders     float64
order_frequency                 object
dtype: object

**NOTE**: key type on both DF are the user ID

In [28]:
#dropping 'eval_set' and '_merge' columns

ords_prods_grouped = ords_prods_grouped.drop(columns=['_merge','eval_set'])

In [ ]:
# Merge ords_prods_grouped with customers data

ords_prods_grouped_merge = ords_prods_grouped.merge(customers, on = 'user_id', indicator = True)

# QUESTION 8. EXPORT NEW DF AS PICKLE

In [ ]:
#Export ords_prods_cust_agg as pickle

ords_prods_grouped_merge.to_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_cust_agg.pkl'))